src/helpers/gateway/mod.rs

```diff
@@ -0,0 +1,158 @@
+mod receive;
+mod send;
+mod transport;
+mod wrapper;
+
+pub use receive::ReceivingEnd;
+pub use send::SendingEnd;
+
+use std::{fmt::Debug, num::NonZeroUsize};
+
+use crate::{
+    helpers::{transport::TransportImpl, ChannelId, Message, Role, RoleAssignment, TotalRecords},
+    protocol::QueryId,
+};
+
+use crate::helpers::gateway::{
+    receive::GatewayReceivers, send::GatewaySenders, transport::RoleResolvingTransport,
+};
+#[cfg(all(feature = "shuttle", test))]
+use shuttle::future as tokio;
+
+/// Gateway into IPA Infrastructure systems. This object allows sending and receiving messages
+pub struct Gateway {
+    config: GatewayConfig,
+    transport: RoleResolvingTransport<TransportImpl>,
+    senders: GatewaySenders,
+    receivers: GatewayReceivers<TransportImpl>,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct GatewayConfig {
+    /// The maximum number of items that can be outstanding for sending.
+    pub send_outstanding: NonZeroUsize,
+    /// The maximum number of items that can be outstanding for receiving.
+    pub recv_outstanding: NonZeroUsize,
+}
+
+impl Gateway {
+    #[must_use]
+    pub fn new(
+        query_id: QueryId,
+        config: GatewayConfig,
+        roles: RoleAssignment,
+        transport: TransportImpl,
+    ) -> Self {
+        Self {
+            config,
+            transport: RoleResolvingTransport {
+                query_id,
+                roles,
+                inner: transport,
+                config,
+            },
+            senders: GatewaySenders::default(),
+            receivers: GatewayReceivers::default(),
+        }
+    }
+
+    #[must_use]
+    pub fn role(&self) -> Role {
+        self.transport.role()
+    }
+
+    #[must_use]
+    pub fn get_sender<M: Message>(
+        &self,
+        channel_id: &ChannelId,
+        total_records: TotalRecords,
+    ) -> SendingEnd<M> {
+        let (tx, maybe_recv) = self
+            .senders
+            .get_or_create(channel_id, self.config.send_outstanding);
+        // I don't understand why rustc complains about unused variables there
+        #[allow(unused)]
```

Comment:
> I think it is because TransportImpl is an enum with no variants when the test-fixture feature is not set.

but we are not using it directly here, `send` is called on `RoleResolvingTransport`. Besides `TransportImpl` defines `send` so I am still puzzled why we have unused warning here

good point about the `Result` though - i missed it . Fixed

Reply:
however `rustc` agrees with you

```bash
 cargo build --release --features test-fixture
```

gives me no warnings, but I don't understand why :)